### PR TITLE
[TEST] Enforce skip headers when needed

### LINKED ISSUE #34650: Improve rest-api-spec test runner to fail tests that use headers but don't list them in the skip section
I'm raising this issue after submitting a [pull request](https://github.com/elastic/elasticsearch/pull/34551) to add `headers` to the `skip` section of some tests that were missing it. I'm requesting that the Elasticsearch test runner fail tests that use features--specifically `headers`-- but don't include those features in the `skip` section.

I became aware of this issue when the Ruby client tests were failing against the latest `rest-api-spec` tests because some tests used headers but didn't specify `features: headers` in the `skip` section. The Ruby client doesn't support headers and wasn't able to properly skip these tests.

### PR DESCRIPTION
**Pull request by javanna**
The java yaml test runner supports sending request headers, yet not all clients support headers. This commit makes sure that we enforce adding a skip section with feature "headers" whenever headers are used in a do section as part of a test. That decreases the chance for new tests to break client builds due to the missing skip section.

Closes #34650

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSection.java [hunk 1]
```diff
@@ -98,6 +98,12 @@ public void addExecutableSection(ExecutableSection executableSection) {
                     + "[skip] so runners that do not support the [node_selector] section can skip the test at line ["
                     + doSection.getLocation().lineNumber + "]");
             }
+            if (false == doSection.getApiCallSection().getHeaders().isEmpty()
+                    && false == skipSection.getFeatures().contains("headers")) {
+                throw new IllegalArgumentException("Attempted to add a [do] with a [headers] section without a corresponding "
+                    + "[skip] so runners that do not support the [headers] section can skip the test at line ["
```
Comment (atorok) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSection.java]: Maybe worth mentioning the feature has the same name to save people some digging ?
```suggestion
                    + "[skip: {"feature": "headers"}] so runners that do not support the [headers] section can skip the test at line ["
```

**Review by alpar-t — APPROVED** [review]
LGTM just a suggestion.

**Comment by javanna** [comment]
retest this please

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSection.java [hunk 2]
```diff
@@ -88,14 +88,21 @@ public void addExecutableSection(ExecutableSection executableSection) {
...
                     && false == skipSection.getFeatures().contains("node_selector")) {
                 throw new IllegalArgumentException("Attempted to add a [do] with a [node_selector] section without a corresponding "
-                    + "[skip] so runners that do not support the [node_selector] section can skip the test at line ["
+                    + "[skip: \"features\": \"node_selector\"] so runners that do not support the [node_selector] section can skip the " +
+                    "test at line ["
+                    + doSection.getLocation().lineNumber + "]");
+            }
+            if (false == doSection.getApiCallSection().getHeaders().isEmpty()
+                    && false == skipSection.getFeatures().contains("headers")) {
+                throw new IllegalArgumentException("Attempted to add a [do] with a [headers] section without a corresponding "
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSection.java]: :+1:

**Review by nik9000 — APPROVED** [review]

**Comment by javanna** [comment]
@alpar or @nik9000 can you maybe have another look? I had to move the validation code higher-up, as this change made another bug surface: we don't look at the skip section from setup or teardown in our validation, and we have quite some skip headers in our setup sections, so false positives without adapting the validation code. Let me know what you think.

**Review by nik9000 — APPROVED** [review]

**Comment by javanna** [comment]
retest this please

**Pull request merged**